### PR TITLE
Add unsafe dma api

### LINKED ISSUE #1512: Have a Move-In-Move-Out DMA transaction API in addition (again)
In #1238 using DMA was made easier but it turned out that the new API doesn't support all use-cases supported by the previous API.

Especially something like https://github.com/georgik/esp-display-interface-spi-dma isn't possible this way.

(Using Polonius partially helps a bit but still doesn't fully solve the issue)

### PR DESCRIPTION
**Pull request by bjoernQ**
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/CONTRIBUTING.md) and followed its instructions.

#### Description
Fixes #1512

Bringing back the old clumsy API doesn't seem to be a good solution so this adds unsafe versions of write_dma, read_dma and transfer_dma for SPI master

#### Testing
Tested with https://github.com/bjoernQ/rust-esp32s3-ili9341 and the existing examples

### CODE REVIEW

File: esp-hal/src/spi/master.rs [hunk 1]
```diff
@@ -1177,7 +1240,21 @@ pub mod dma {
                     &mut self.channel.rx,
                 )?;
             }
-            Ok(SpiDmaTransferRxTx { spi_dma: self })
+
+            Ok(())
+        }
+
+        /// Wait for the DMA transfer to complete
+        pub fn wait(&mut self) -> Result<(), DmaError> {
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply esp-hal/src/spi/master.rs:1248]: Ah flush is blocking, okay this makes sense now. Will flush ever get stuck if a DMA transaction fails?

[the remark this answers is not in the document]
Comment (bjoernQ) [reply esp-hal/src/spi/master.rs:1248]: I tried to get it into an error state and it didn't get stuck. Also, it really seems to be fine to just check `SPI::cmd::usr`

**Review by JurajSadel — APPROVED** [review]
LGTM, thanks!

**Review by MabezDev — CHANGES_REQUESTED** [review]
I'm not against having this as an API, but I think we should explore what we're trying to solve in https://github.com/esp-rs/esp-hal/issues/1512 first.

**Comment by jessebraham** [comment]
Marking this as draft for now.

**Comment by bjoernQ** [comment]
> Marking this as draft for now.

Thanks for reminding me of this. I will close this since we should find a better solution than going unsafe

**Pull request closed without merging**